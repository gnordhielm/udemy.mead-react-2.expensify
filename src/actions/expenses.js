import db from '../firebase/firebase'
import uuid from 'uuid'

// Expenses actions

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData={}) => {

  return dispatch => {

    const {
      description='',
      note='',
      amount=0,
      createdAt=0
    } = expenseData

    const expense = {
      description, note, amount, createdAt
    }

    return db.ref('expenses')
      .push(expense)
      .then(ref => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })

  }
}

export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
})


export const updateExpense = ({ id, changes }) => ({
  type: 'UPDATE_EXPENSE',
  id, changes
})
