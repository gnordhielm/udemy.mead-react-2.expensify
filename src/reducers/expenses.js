
const initialState = []

export default (state=initialState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ]

    case 'REMOVE_EXPENSE':
      if (!action.id)
        throw new Error('remove expense requires an id')
      return state.filter(({ id }) => id !== action.id)

    case 'UPDATE_EXPENSE':
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.changes
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}
