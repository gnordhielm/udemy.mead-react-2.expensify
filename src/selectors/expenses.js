import moment from 'moment'

const sortByLookup = {
  'date': 'createdAt',
  'amount': 'amount'
}

export const getVisibleExpenses = (expenses, {
  text, sortBy, startDate, endDate
}) => expenses
  .filter(({ description, createdAt }) => {

    const createdAtMoment = moment(createdAt)

    const startDateMatch = startDate ?
      startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ?
      endDate.isSameOrAfter(createdAtMoment, 'day') : true

    const textMatch = description
      .toLowerCase()
      .includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  })
  .sort((a, b) => b[sortByLookup[sortBy]] - a[sortByLookup[sortBy]])

export const getExpensesTotal = expenses => expenses
    .reduce((acc, expense) => acc + expense.amount, 0)
