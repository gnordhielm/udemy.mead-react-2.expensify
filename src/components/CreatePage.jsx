import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm.jsx'
import { startAddExpense } from '../actions/expenses'

export class CreatePage extends React.Component {

  onSubmit = expense => {

    this.props.startAddExpense(expense)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <h3>Add Expense</h3>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    )
  }
}

const mapDispatch = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
})

export default connect(undefined, mapDispatch)(CreatePage)
