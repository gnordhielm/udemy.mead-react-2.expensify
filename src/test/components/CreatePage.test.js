import 'react-dates/initialize'

import React from 'react'
import { shallow } from 'enzyme'
import { CreatePage } from '../../components/CreatePage.jsx'
import expenses from '../fixtures/expenses'

let startAddExpense, history, wrapper

beforeEach(() => {
  startAddExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(<CreatePage
    startAddExpense={startAddExpense}
    history={history}
  />)
})

test('should render CreatePage', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle startAddExpense', () => {
  const newExpense = {
    ...expenses[0], id: undefined
  }

  wrapper.find('ExpenseForm').prop('onSubmit')(newExpense)

  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(startAddExpense).toHaveBeenLastCalledWith(newExpense)

  expect(wrapper).toMatchSnapshot()

})
