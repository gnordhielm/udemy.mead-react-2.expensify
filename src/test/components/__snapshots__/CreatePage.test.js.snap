// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle startAddExpense 1`] = `
<div>
  <h3>
    Add Expense
  </h3>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;

exports[`should render CreatePage 1`] = `
<div>
  <h3>
    Add Expense
  </h3>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
